### PR TITLE
Wikicat: Enforce category_contains constraint

### PR DESCRIPTION
**Pull request by rahul1980**
If a category has "category_contains" populated, then we restrict its members to belong to that class.
e.g. ("american cyclists", "category_contains", "human") and ("Lance Armstrong", "instance of", "human").

The instance_of check is done transitively using subclass-of edges, and if a category doesn't have a category_contains  role, then the check is not performed.

### CODE REVIEW

File: sling/nlp/wikicat/generator.py [hunk 1]
```diff
@@ -294,8 +296,32 @@ def post_process(self, parses):
   # Returns category members that are not categories themselves.
   def get_members(self, frame):
     output = []
+    allowed = []
+    store = frame.store()
+    if self.h_category_contains in frame:
```
Comment (anders-sandholm) [inline sling/nlp/wikicat/generator.py]: Minor thing. Dealing with the entire case of `if self.h_category_contains not in frame:` first (including looping over members and a `return output` at the end) might be easier to read albeit slightly more code.

**Comment by ringgaard** [comment]
Please use "git checkout dev FILES..." instead of "git cherry-pick" when creating the PRs to avoid pulling in a long and irrelevant commit list into the PR.

File: sling/nlp/wikicat/generator.py [hunk 2]
```diff
@@ -294,8 +296,32 @@ def post_process(self, parses):
...
+        for value in member(self.h_instanceof):
+          # Fast-check for whether 'value' satisfies category_contains.
+          if value in allowed:
+            valid = True
+            break
+
+          # Slow-check for whether 'value' satisfies category_contains.
+          for a in allowed:
+            if self.extractor.in_closure(\
+               frame.store(), self.h_subclassof, a, value):
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/nlp/wikicat/generator.py]: Done.

File: sling/nlp/wikicat/generator.py [hunk 3]
```diff
@@ -294,8 +296,32 @@ def post_process(self, parses):
   # Returns category members that are not categories themselves.
   def get_members(self, frame):
     output = []
+    allowed = []
+    store = frame.store()
+    if self.h_category_contains in frame:
+      allowed = [store.resolve(c) for c in frame(self.h_category_contains)]
     for member in frame(self.h_member):
-      if not self.is_category(member):
+      if self.is_category(member):
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/nlp/wikicat/generator.py]: I had considered this earlier but the issue is that super-categories are often very diffused, e.g. "1890s Fashion", which contains the much better "clothing companies established in <year>" categories (year = 1890, 1891, etc). In such cases it doesn't make sense to do this inclusion.

OTOH, there are legit cases like "American cyclists" that contain "American male cyclists". Unfortunately I couldn't see a way to distinguish between the two.

File: sling/nlp/wikicat/generator.py (hunk 1 above)
Comment (rahul1980) [reply sling/nlp/wikicat/generator.py]: Good suggestion. Done.

**Comment by rahul1980** [comment]
> Mostly minor comments.
> LGTM.
> Additional off-topic comment. We could re-write `is_category` (l. 104) to the following one-liner `return self.h_category in frame(self.h_instanceof)`

Done.

**Comment by rahul1980** [comment]
> Please use "git checkout dev FILES..." instead of "git cherry-pick" when creating the PRs to avoid pulling in a long and irrelevant commit list into the PR.

Actually I wasn't doing either. I just happened to reuse an old local branch (that was used to push an earlier PR).

**Comment by ringgaard** [comment]
> > Please use "git checkout dev FILES..." instead of "git cherry-pick" when creating the PRs to avoid pulling in a long and irrelevant commit list into the PR.
> 
> Actually I wasn't doing either. I just happened to reuse an old local branch (that was used to push an earlier PR).

If you are using an old branch, you should make sure it is synced to upstream/master by doing "git fetch upstream master" and "git merge upstream/master".

**Pull request merged**